elyra/contents/file_parser.py

```diff
@@ -0,0 +1,193 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import os
+import nbformat
+import re
+
+from abc import ABC, abstractmethod
+
+from typing import Any, Type, TypeVar, List, Dict
+from traitlets import log
+from ..util import get_expanded_path, get_absolute_path
+
+# Setup forward reference for type hint on return from class factory method.  See
+# https://stackoverflow.com/questions/39205527/can-you-annotate-return-type-when-value-is-instance-of-cls/39205612#39205612
+F = TypeVar('F', bound='FileParser')
+
+
+class FileParser(ABC):
+
+    @classmethod
+    def get_instance(cls: Type[F], **kwargs: Any) -> F:
+        """Creates an appropriate subclass instance based on the extension of the filepath"""
+        filepath = kwargs['filepath']
+
+        if '.ipynb' in filepath:
+            return NotebookFileParser(filepath)
+        elif '.py' in filepath:
+            return PythonFileParser(filepath)
+        elif '.r' in filepath:
+            return RFileParser(filepath)
+        else:
+            raise ValueError('Unsupported file type: {}'.format(filepath))
+
+    _type = None
+
+    def __init__(self, operation_filepath):
+        self._operation_filepath = operation_filepath
+        self.parser = None
+
+        # Consider removing or add log messages
+        self.log = log.get_logger()
+
+    @property
+    @abstractmethod
+    def type(self):
+        pass
+
+    @property
+    def operation_filepath(self):
+        # Retrieve the absolute path
+        root_dir = get_expanded_path()
+        abs_path = get_absolute_path(root_dir, self._operation_filepath)
+
+        if not os.path.exists(abs_path):
+            raise FileNotFoundError(f'Could not find {abs_path}')
+        if not os.path.isfile(abs_path):
+            raise ValueError(f'Not a file: {abs_path}')
+
+        self._operation_filepath = abs_path
+        return self._operation_filepath
+
+    @abstractmethod
+    def get_next_code_chunk(self) -> List[str]:
+        # Implements a generator for lines of code for the specified operation type
+        with open(self.operation_filepath) as f:
+            for line in f:
+                yield [line.strip()]
+
+    def get_resources(self):
+        # Returns a model dictionary of all the regex matches for each key in the regex dictionary
+        model = dict()
+        model["env_list"] = dict()
+        model["inputs"] = dict()
+        model["outputs"] = dict()
+
+        for chunk in self.get_next_code_chunk():
+            if chunk:
+                for line in chunk:
+                    if not self.parser:
+                        raise RuntimeError(f'Could not find appropriate language parser for {self.operation_filepath}')
```

Comment:
This should be moved, to the first thing this method checks.  Or, perhaps better, make `parser` a property, since it should have a value following construction - then let the accessor throw a `ValueError` if it finds it not set.

Reply:
Ugh, of course! No idea why I put it there yesterday ha